Support powershell core

twangboy opened the pull request:
### What does this PR do?
Adds support for powershell core. The binary for powershell core is `pwsh`. Added tests.
Fixes running encoded powershell commands. This broke in Python 3, but we didn't know it. Added tests.
Moves a couple integration tests to pytest/functional

Powershell core is added to the golden image in this PR: https://github.com/saltstack/salt-jenkins/pull/1688

When present, tests will run for pwsh binaries

### What issues does this PR fix or reference?
Fixes: https://github.com/saltstack/salt/issues/58598

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**
<!-- Please review the [test documentation](https://docs.saltstack.com/en/master/topics/tutorials/writing_tests.html) for details on how to implement tests into Salt's test suite. -->
- [x] Docs
- [x] Changelog - https://docs.saltstack.com/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
Yes

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

twangboy committed:
Add support for powershell 7 to cmd module

twangboy committed:
Add tests, fix tests

twangboy committed:
Add changelog

twangboy committed:
Remove debug code. Fix tests

s0undt3ch commented on salt/modules/cmdmod.py:
```diff
@@ -354,15 +358,11 @@ def _run(
-            cmd = "Powershell -NonInteractive -NoProfile -ExecutionPolicy Bypass {}".format(
-                cmd.replace('"', '\\"')
-            )
+            cmd = f'"{shell}" -NonInteractive -NoProfile -ExecutionPolicy Bypass -Command {cmd}'
```
Hmm. I know we run Py3.6+ on windows packages, actually 3.7, but salt itself still supports 3.5, which doesn't support f-strings....

twangboy replied:
You're right... coming from SSE, I forget we're still on 3.5

s0undt3ch commented on salt/modules/cmdmod.py:
```diff
@@ -3663,6 +3663,11 @@ def powershell(
+    if shell not in ["powershell", "pwsh"]:
+        msg = "Must specify a valid powershell binary. Must be 'powershell' " \
+              "or 'pwsh'"
+        raise CommandExecutionError(msg)
```
Why not `raise CommandExecutionError("The Message")`?
Why create a `msg` variable for a single use?

twangboy replied:
Done

s0undt3ch commented on tests/pytests/functional/modules/test_cmdmod.py:
```diff
@@ -0,0 +1,123 @@
+import salt.utils.path
+
+pytestmark = [
+    pytest.mark.windows_whitelisted
```
If you add a trailing comma, black won't reformat this code chunk

twangboy replied:
Done

s0undt3ch commented on tests/pytests/functional/modules/test_cmdmod.py:
```diff
@@ -0,0 +1,123 @@
+import salt.utils.path
+
+pytestmark = [
+    pytest.mark.windows_whitelisted
```
```suggestion
    pytest.mark.windows_whitelisted,
    pytest.mark.skip_unless_on_windows,
```

twangboy replied:
Done. Though if we add non Windows tests here in the future, we'll have to move them back.

s0undt3ch commented on tests/pytests/functional/modules/test_cmdmod.py:
```diff
@@ -0,0 +1,123 @@
+    assert ret == 'foo'
+
+
+@pytest.mark.skip_unless_on_windows(reason="Only run on Windows")
```
I suggested a change in the above `pytestmark` so that you don't have to add this decorator to all test functions.

twangboy replied:
Done

s0undt3ch commented on tests/pytests/functional/modules/test_cmdmod.py:
```diff
@@ -0,0 +1,123 @@
+
+
+@pytest.mark.skip_unless_on_windows(reason="Only run on Windows")
+@pytest.mark.skipif(not salt.utils.path.which("pwsh"), reason="Powershell 7 Not Present")
```
```suggestion
@pytest.mark.skip_if_binaries_missing("pwsh", message="Powershell 7 Not Present")
```

twangboy replied:
Thanks... that's my ignorance in not knowing the names of all the skip things...

s0undt3ch reviewed (changes requested):
My only real concern are `f-strings` but, if Debian 9 and Ubuntu 16.04 don't choke, I'm cool with it.

s0undt3ch commented:
Yep. Debian 9 choked. We can't, **yet**, use `f-strings`

max-arnold commented:
Does it fix these issues?

https://github.com/saltstack/salt/pull/31351#issuecomment-561042603
https://github.com/saltstack/salt/pull/31351#issuecomment-561046410

twangboy commented:
@max-arnold Yeah, I think so... I'll add some tests around your example

twangboy committed:
Improved docs for encoded_cmd in cmd.run and cmd.run_all

twangboy commented:
@max-arnold Your first comment using `cmd.powershell` with `encode_cmd=True` works. I fixed the issue with the 2nd comment in the `cmd.powershell` function by prepending `$ProgressPreference='SilentlyContinue'` to the command before it is encoded.

With `cmd.run` and `cmd.run_all` however, that is impossible as the command is being passed to the function already encoded. So, I added some documentation about the need to prepend `$ProgressPreference='SilentlyContinue'` to the command before you encode it if you want to remove the raw xml data in the return.

As a note, this seems to only affect older versions of powershell. I don't get the raw xml data with Powershell 7, so maybe they've addressed that issue there.

twangboy committed:
Fix some lint

twangboy committed:
Fix some more lint

xeacott commented on tests/integration/modules/test_cmdmod.py:
```diff
@@ -557,38 +552,27 @@ def test_windows_env_handling(self):
-            "cmd.run_all", cmd="echo salt", python_shell=False, shell="powershell"
-        )
-        self.assertEqual(out["stdout"], "salt")
+        val = "i like cheese"
```
But do you?

s0undt3ch replied:
Melted.

s0undt3ch committed:
Less code same tests

twangboy committed:
Add docs to decorated function

twangboy committed:
Remove trailing comma from kwargs

twangboy committed:
Revert "Remove trailing comma from kwargs"

This reverts commit 225328484b22dda769257c89eedb6a60129272b5.

twangboy committed:
Revert "Remove trailing comma from kwargs"

This reverts commit 225328484b22dda769257c89eedb6a60129272b5.

twangboy committed:
Revert "Remove trailing comma from kwargs"

This reverts commit 225328484b22dda769257c89eedb6a60129272b5.

twangboy committed:
Revert "Remove trailing comma from kwargs"

This reverts commit 225328484b22dda769257c89eedb6a60129272b5.

twangboy committed:
Specify powershell core instead of 7

twangboy committed:
Use f string on Windows

twangboy committed:
Use f string on Windows

twangboy committed:
Use f string on Windows

twangboy committed:
Use f string on Windows

twangboy committed:
Fix slowTest marker

twangboy committed:
Don't use f strings

twangboy committed:
Fix pre-commit

twangboy committed:
Remove comma after kwargs

twangboy committed:
Remove comma after kwargs

twangboy committed:
Remove comma after kwargs

twangboy committed:
Remove comma after kwargs

twangboy committed:
Don't use fstrings

twangboy committed:
Remove fstrings in test

twangboy committed:
Remove fstrings in test

twangboy committed:
Remove fstrings in test

twangboy committed:
Remove fstrings in test

twangboy committed:
Fix pre-commit

twangboy committed:
Fix pre-commit

twangboy committed:
Fix pre-commit

twangboy committed:
Fix pre-commit

s0undt3ch commented on tests/pytests/functional/modules/test_cmdmod.py:
```diff
@@ -0,0 +1,104 @@
+import base64
```
Let's rename this to `tests/pytests/functional/modules/cmdmod/test_powershell.py `

s0undt3ch commented on tests/integration/modules/test_cmdmod.py:
```diff
@@ -546,38 +546,31 @@ def test_windows_env_handling(self):
     @pytest.mark.slow_test
     @skipIf(not salt.utils.platform.is_windows(), "minion is not windows")
-    def test_windows_powershell_script_args(self):
+    @skipIf(not salt.utils.path.which("pwsh"), "Requires Powershell 7")
```
```suggestion
    @pytest.mark.skip_if_binaries_missing("pwsh", reason="Requires Powershell 7")
```

twangboy committed:
Use pytest skips

twangboy committed:
Add message to skips

twangboy commented:
Not sure why Fedora 33 is failing... Everything else is passing

twangboy committed:
Add missing import
